### PR TITLE
Update to mypy 0.971

### PR DESCRIPTION
**Pull request by wence-**
## Description

Update to current mypy, primarily since on Apple silicon, the previous pinned version of mypy is no longer installable (`typed-ast` does not build from source). This necessitates some minor updates to the type annotation rules, since the newer mypy version is a bit pickier.

While we're here, exclude directories we were previously just ignoring errors in.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (bdice) [inline .pre-commit-config.yaml:32]: Can you explain why `--namespace-packages --explicit-package-bases` are needed here? https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-namespace-packages

File: setup.cfg
Comment (vyasr) [inline setup.cfg:43]: Why do we need to set those options now if it's causing this extra difficulty?

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:15]: I'm fine with this change, but I'm curious: do newer versions of mypy now complain about this?

**Review by bdice — COMMENTED** [review]
We'll also want to update the conda environment: https://github.com/rapidsai/cudf/blob/8ad0290b7f57eb7df903b855666c2d8cd712a641/conda/environments/cudf_dev_cuda11.5.yml#L46.

I expect this may cause compatibility issues and require a broader update of mypy across RAPIDS. Alternatively, we can remove mypy from the conda environment (so it can't conflict with other RAPIDS envs) and solely rely on pre-commit for managing the version and executing the check.

**Review by vyasr — COMMENTED** [review]
This is a surprisingly minimal changeset. I thought the update required more changes than this, but perhaps that was due to unrelated incompatibilities? If CI passes then that's good enough for me.

**Comment by vyasr** [comment]
> We'll also want to update the conda environment:
> 
> https://github.com/rapidsai/cudf/blob/8ad0290b7f57eb7df903b855666c2d8cd712a641/conda/environments/cudf_dev_cuda11.5.yml#L46
> 
> I expect this may cause compatibility issues and require a broader update of mypy across RAPIDS. Alternatively, we can remove mypy from the conda environment (so it can't conflict with other RAPIDS envs) and solely rely on pre-commit for managing the version and executing the check.

I believe cudf is the only package in RAPIDS that uses mypy at all, so it shouldn't cause any problems (unless there are incompatibilities with dependencies).

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py:15]: Ah here's the issue. `NotImplementedType` was only [introduced in Python 3.10](https://docs.python.org/3/library/types.html#types.NotImplementedType), so we can't use that here.

File: .pre-commit-config.yaml
Comment (wence-) [reply .pre-commit-config.yaml:32]: The `test` and `benchmarks` directories don't contain `__init__.py` and so mypy complains about multiple `conftest` modules:
```
$ mypy --config-file=setup.cfg python/cudf
python/cudf/benchmarks/conftest.py: error: Duplicate module named "conftest" (also at "python/cudf/benchmarks/internal/conftest.py")
python/cudf/benchmarks/conftest.py: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more info
python/cudf/benchmarks/conftest.py: note: Common resolutions include: a) using `--exclude` to avoid checking one of them, b) adding `__init__.py` somewhere, c) using `--explicit-package-bases` or adjusting MYPYPATH
Found 1 error in 1 file (errors prevented further checking)
```

File: setup.cfg
Comment (wence-) [reply setup.cfg:43]: https://github.com/rapidsai/cudf/pull/11640#discussion_r960878090

File: python/cudf/cudf/core/dataframe.py
Comment (wence-) [reply python/cudf/cudf/core/dataframe.py:15]: Ah, let me check in 3.8 and 3.9

File: .pre-commit-config.yaml
Comment (vyasr) [reply .pre-commit-config.yaml:32]: Ah yes I remember seeing this error before. My original plan was to address this by moving the tests/benchmarks outside the main repo, but that's a bigger task that requires more coordination. I'm fine with this solution as a stopgap.

File: python/cudf/cudf/core/dataframe.py
Comment (wence-) [reply python/cudf/cudf/core/dataframe.py]: Added a workaround

File: .pre-commit-config.yaml
Comment (bdice) [reply .pre-commit-config.yaml:33]: We should just `--exclude` tests and benchmarks, right? We don't need to run mypy on those.

File: python/cudf/cudf/core/dataframe.py
Comment (wence-) [reply python/cudf/cudf/core/dataframe.py]: > I'm fine with this change, but I'm curious: do newer versions of mypy now complain about this?

Yes: https://github.com/python/mypy/issues/4791

I've backwards-compated by setting `NotImplementedType = Any` because I can't see another way of doing things.

File: .pre-commit-config.yaml
Comment (wence-) [reply .pre-commit-config.yaml:33]: We ignore all the errors in the tests subdir, so I guess we could do that.

Comment (wence-) [reply .pre-commit-config.yaml]: Done in 114b4b1bdf

File: setup.cfg
Comment (wence-) [inline setup.cfg:41]: I'm now `--exclude`ing these modules, rather than parsing and then ignoring errors.

File: .pre-commit-config.yaml
Comment (bdice) [inline .pre-commit-config.yaml:34]: Can these be specified in the `setup.cfg` instead of the pre-commit flags? https://mypy.readthedocs.io/en/stable/config_file.html#confval-exclude

That way it will be more consistent if mypy is run outside of pre-commit.

File: python/cudf/cudf/core/column/string.py
Comment (bdice) [inline python/cudf/cudf/core/column/string.py:3654]: Can we use a “cast” instead of ignoring the type?

File: python/cudf/cudf/core/dataframe.py
Comment (bdice) [inline python/cudf/cudf/core/dataframe.py:106]: Rather than hard-code the version, try to import it and define the fallback in an exception handler.

```python
try:
    # Introduced in Python 3.10
    from types import NotImplementedType
except ImportError:
    NotImplementedType = Any
```

File: python/cudf/cudf/core/index.py
Comment (bdice) [inline python/cudf/cudf/core/index.py:234]: Can we give this a return type like `def … () -> Union[…]:`?

File: python/cudf/cudf/core/indexed_frame.py
Comment (bdice) [inline python/cudf/cudf/core/indexed_frame.py:12]: Needs to be guarded with a try/except as above.

File: python/cudf/cudf/core/multiindex.py
Comment (bdice) [inline python/cudf/cudf/core/multiindex.py:1463]: Instead of ignoring the type:
```suggestion
    def is_unique(self) -> bool:
```

File: setup.cfg
Comment (bdice) [reply setup.cfg:41]: Smart. Should we do the same for the _version files?

File: .pre-commit-config.yaml
Comment (wence-) [reply .pre-commit-config.yaml]: Done.

File: setup.cfg
Comment (wence-) [reply setup.cfg:41]: Good idea, done.

File: python/cudf/cudf/core/dataframe.py
Comment (wence-) [reply python/cudf/cudf/core/dataframe.py:106]: Unfortunately that doesn't work (because `ImportError` is not a statically determinable condition):

In [3.8 or 3.9](https://mypy-play.net/?mypy=latest&python=3.8&gist=181adc25354058b0dc50790c1934a78f)
```
main.py:3: error: Module "types" has no attribute "NotImplementedType"
Found 1 error in 1 file (checked 1 source file)
```

In [3.10 or 3.11](https://mypy-play.net/?mypy=latest&python=3.10&gist=181adc25354058b0dc50790c1934a78f)
```
main.py:5: error: Cannot assign multiple types to name "NotImplementedType" without an explicit "Type[...]" annotation
main.py:5: error: Incompatible types in assignment (expression has type "object", variable has type "Type[_NotImplementedType]")
Found 2 errors in 1 file (checked 1 source file)
```

File: python/cudf/cudf/core/index.py
Comment (wence-) [reply python/cudf/cudf/core/index.py:234]: That doesn't help:

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

python/cudf/cudf/core/index.py:234: error: Decorated property not supported
```

File: python/cudf/cudf/core/multiindex.py
Comment (wence-) [reply python/cudf/cudf/core/multiindex.py:1463]: As above, this doesn't help with the "decorated property not supported" error

File: python/cudf/cudf/core/dataframe.py
Comment (wence-) [reply python/cudf/cudf/core/dataframe.py]: Yeah, we need a `sys.version` gated import which is now done.

File: python/cudf/cudf/core/column/string.py
Comment (wence-) [reply python/cudf/cudf/core/column/string.py:3655]: Done.

File: .pre-commit-config.yaml
Comment (wence-) [reply .pre-commit-config.yaml]: Note that if you run mypy yourself you get errors that the mypy in pre-commit doesn't see (because your environment has numpy installed so mypy has more type annotations in that case and gets confused).

Comment (bdice) [reply .pre-commit-config.yaml]: In that case we should add the appropriate typesheds to the pre-commit extras.

Comment (bdice) [reply .pre-commit-config.yaml]: Example from another project: https://github.com/glotzerlab/signac/blob/3d546da97da810b55ce7d03028b692d8dd743b23/.pre-commit-config.yaml#L59

File: python/cudf/cudf/core/dataframe.py
Comment (bdice) [reply python/cudf/cudf/core/dataframe.py]: Wow! I had not anticipated that static condition. That’s interesting.

File: python/cudf/cudf/core/index.py
Comment (bdice) [reply python/cudf/cudf/core/index.py:234]: Okay, here is a second question. Can the `# type: ignore` comment be on the same line as `def` and not the property decorator?

Comment (wence-) [reply python/cudf/cudf/core/index.py:234]: No. It needs to be on the property decorator line. It's also a terrible warning, since mypy can actually deduce the type (this appears like it was fixed recently) https://github.com/python/mypy/issues/1362

File: .pre-commit-config.yaml
Comment (wence-) [reply .pre-commit-config.yaml]: The issue (is it an issue?) here is that numpy typestubs are now _in_ numpy (so the only way to get them is to install numpy in the pre-commit venv). If that is reasonable, this is the additional diff I need. It's mostly working around the issue that untagged union types are arguably an anti-pattern, but python has no support for tagged unions and so this is where we are

```patch
diff --git a/python/cudf/cudf/core/column/column.py b/python/cudf/cudf/core/column/column.py
index 194377a7c9..0b963c5506 100644
--- a/python/cudf/cudf/core/column/column.py
+++ b/python/cudf/cudf/core/column/column.py
@@ -1057,7 +1057,7 @@ class ColumnBase(Column, Serializable, BinaryOperand, Reducible):
         frames = []
         header["type-serialized"] = pickle.dumps(type(self))
         try:
-            dtype, dtype_frames = self.dtype.serialize()
+            dtype, dtype_frames = self.dtype.serialize()  # type: ignore[union-attr]
             header["dtype"] = dtype
             frames.extend(dtype_frames)
             header["dtype-is-cudf-serialized"] = True
diff --git a/python/cudf/cudf/core/column/numerical.py b/python/cudf/cudf/core/column/numerical.py
index 4b74dde129..7fc4e8b81a 100644
--- a/python/cudf/cudf/core/column/numerical.py
+++ b/python/cudf/cudf/core/column/numerical.py
@@ -591,11 +591,13 @@ class NumericalColumn(NumericalBaseColumn):
             else:
                 # Kinds are the same but to_dtype is smaller
                 if "float" in to_dtype.name:
-                    finfo = np.finfo(to_dtype)
+                    finfo: Any = np.finfo(to_dtype)
                     lower_, upper_ = finfo.min, finfo.max
                 elif "int" in to_dtype.name:
-                    iinfo = np.iinfo(to_dtype)
+                    iinfo: Any = np.iinfo(to_dtype)
                     lower_, upper_ = iinfo.min, iinfo.max
+                else:
+                    raise RuntimeError("Unhandled kind")
 
                 if self.dtype.kind == "f":
                     # Exclude 'np.inf', '-np.inf'
@@ -634,7 +636,7 @@ class NumericalColumn(NumericalBaseColumn):
 
         # want to cast int to float
         elif self.dtype.kind in {"i", "u"} and to_dtype.kind == "f":
-            info = np.finfo(to_dtype)
+            info: Any = np.finfo(to_dtype)
             biggest_exact_int = 2 ** (info.nmant + 1)
             if (self.min() >= -biggest_exact_int) and (
                 self.max() <= biggest_exact_int
diff --git a/python/cudf/cudf/core/column/timedelta.py b/python/cudf/cudf/core/column/timedelta.py
index e6d688014f..c496482d38 100644
--- a/python/cudf/cudf/core/column/timedelta.py
+++ b/python/cudf/cudf/core/column/timedelta.py
@@ -3,7 +3,7 @@
 from __future__ import annotations
 
 import datetime
-from typing import Any, Sequence, cast
+from typing import Any, Sequence, Union, cast
 
 import numpy as np
 import pandas as pd
@@ -164,7 +164,7 @@ class TimeDeltaColumn(ColumnBase):
             return NotImplemented
 
         this: ColumnBinaryOperand = self
-        out_dtype = None
+        out_dtype: Any = None
 
         if is_timedelta64_dtype(other.dtype):
             # TODO: pandas will allow these operators to work but return false
@@ -224,6 +224,7 @@ class TimeDeltaColumn(ColumnBase):
             if np.isnat(other):
                 return cudf.Scalar(None, dtype=self.dtype)
 
+            common_dtype: Union[str, Dtype]
             if other_time_unit not in {"s", "ms", "ns", "us"}:
                 common_dtype = "timedelta64[s]"
             else:
diff --git a/python/cudf/cudf/core/join/_join_helpers.py b/python/cudf/cudf/core/join/_join_helpers.py
index 790f295199..2589c29fad 100644
--- a/python/cudf/cudf/core/join/_join_helpers.py
+++ b/python/cudf/cudf/core/join/_join_helpers.py
@@ -72,9 +72,9 @@ def _match_join_keys(
         if left_is_categorical:
             if how in {"left", "leftsemi", "leftanti"}:
                 return lcol, rcol.astype(ltype)
-            common_type = ltype.categories.dtype
+            common_type = cast(CategoricalDtype, ltype).categories.dtype
         else:
-            common_type = rtype.categories.dtype
+            common_type = cast(CategoricalDtype, rtype).categories.dtype
         return lcol.astype(common_type), rcol.astype(common_type)
 
     if is_dtype_equal(ltype, rtype):
```

I'm -epsilon on adding numpy to the mypy environment, because it's not typed well enough, and I think will create more friction than is necessary.

**Comment by bdice** [comment]
@wence- Before merging, can you please write a PR description describing the motivation for this change and the scope of the changes? (Ideally we would use the PR template checklist, too.)

File: .pre-commit-config.yaml
Comment (bdice) [reply .pre-commit-config.yaml]: Hmm. This could go either way -- my preference would be to have the results of `pre-commit run mypy [--all-files]` align with the results of running `mypy` directly. However, the introduction of a NumPy dependency in the pre-commit environment is less than ideal. I defer to your judgement on this, or perhaps @vyasr if you'd like to weigh in.

File: conda/environments/cudf_dev_cuda11.5.yml
Comment (bdice) [inline conda/environments/cudf_dev_cuda11.5.yml:47]: I'd leave this unpinned. It's going to go out of date too quickly, and we left `cachetools` unpinned in the conda environment below.
```suggestion
  - types-cachetools
```

File: .pre-commit-config.yaml
Comment (bdice) [inline .pre-commit-config.yaml:32]: I'd leave this unpinned. It's going to go out of date too quickly, and we left `cachetools` unpinned in the corresponding conda environment.
```suggestion
                additional_dependencies: [types-cachetools]
```

File: setup.cfg
Comment (bdice) [inline setup.cfg:68]: TIL we don't have any `cudf_kafka` tests. https://github.com/rapidsai/cudf/tree/branch-22.10/python/cudf_kafka/cudf_kafka

**Review by bdice — APPROVED** [review]

File: .pre-commit-config.yaml
Comment (wence-) [reply .pre-commit-config.yaml]: I think the problem with that is where do you stop? pandas also maintain type annotations and type-stubs within pandas, rather than spun out, so do we also need pandas in the pre-commit environment, and then...

I don't think there's a nice way to solve this. For something like `clang-format` via `pre-commit` all we ask for is the binary, and rely on the dev environment to provide the headers. In contrast, installing a sandboxed `mypy` needs us to inject the conda dev environment's `sys.path` into `MYPYPATH` to get consistency between `mypy` and `pre-commit run mypy`.

**Comment by wence-** [comment]
> @wence- Before merging, can you please write a PR description describing the motivation for this change and the scope of the changes? (Ideally we would use the PR template checklist, too.)

Done.

**Comment by wence-** [comment]
Final question to address is what we want to do about missing imports in the pre-commit environment.

File: .pre-commit-config.yaml
Comment (bdice) [reply .pre-commit-config.yaml]: > where do you stop?

You would install all packages from [this list](https://github.com/rapidsai/cudf/blob/branch-22.10/conda/environments/cudf_dev_cuda11.5.yml) that have typing stubs that we care about validating against. To the best of my knowledge, that list would likely include numpy and pandas -- I'd bet we would be happy with only those two packages.

I don't quite follow the discussion of `clang-format` and injecting a conda environment path for mypy. We certainly don't want to mix a user's conda environment with the pre-commit managed environments. My intent was to use `additional_dependencies` to specify a small number of additional packages that should be used solely for typing stubs such that `mypy` in a user conda environment built from the yml linked above would act similarly to `pre-commit run mypy`.

I am happy to defer this conversation for later, since it can be handled separately from the main goal of updating to mypy 0.971.

Comment (wence-) [reply .pre-commit-config.yaml]: > I don't quite follow the discussion of `clang-format` and injecting a conda environment path for mypy. 

Ah, sorry, I was thinking `clangd` editor type-checking. Of course, there's no need to have linting based type-checking for the C++ code (since if it compiles it type-checks). I think what I meant was that if the developer configures their editor to use live type hints they will probably not do so using the `pre-commit` provided mypy (because you can't figure out the path to it programmatically), so they'll just use a version that's installed in the development environment which may pick up other packages.

> I don't know of cudf relying on typing stubs from cupy or dask or anything else.

While cudf may not _rely_ on them, the problem appears as soon as the module is importable and provides types to mypy. Where previously a function would have deduced return type `Any` (which always provides any attributes), it will now have a narrower type which might mean type-checking fails.

> I am happy to defer this conversation for later, since it can be handled separately from the main goal of updating to mypy 0.971.

Yeah, let's do that separately (https://github.com/rapidsai/cudf/issues/11661)

**Review by Ethyling — APPROVED** [review]

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]
This will also need to be updated in the `integration` repository below. Can you update the `mypy` version and add the `types-cachetools` package to the `rapids-build-env` package?

- https://github.com/rapidsai/integration/blob/branch-22.10/conda/recipes/versions.yaml#L103-L104
- https://github.com/rapidsai/integration/blob/branch-22.10/conda/recipes/rapids-build-env/meta.yaml

**Comment by bdice** [comment]
@ajschmidt8 @wence- @vyasr Longer term, as the integration repo is phased out, we may not want to keep “developer tools” like mypy, flake8, black in the conda environment if they are being managed by pre-commit for a given repo. Users wanting consistency with CI style checks must run mypy through pre-commit and not as a standalone executable from the conda environment. There’s not much reason to keep mypy in the conda environment unless it helps developers’ IDE settings, but the results will not necessarily be consistent with what pre-commit’s mypy expects. The pre-commit environment is ultimately our source of truth for CI and local pre-commit checks.

**Comment by ajschmidt8** [comment]
@bdice, so pre-commit is also used in CI? If that's the case, then I'm fine with either of the following:

1. Removing `mypy` from our `integration` repository entirely
2. Updating it to match the version in this PR

Based on your comment, it seems like it never really needed to be in the `integration` repository packages in the first place, right?

**Comment by bdice** [comment]
@ajschmidt8 We used the integration repo's copy of `mypy` until https://github.com/rapidsai/cudf/pull/9412. That changed cudf over to using the `pre-commit` executable (and its managed versions of various linters/formatters) instead of the versions of `black`, `flake8`, `mypy`, etc. from the user's environment (which is the `integration` environment on CI). To my knowledge, the _only_ reason we should keep those tools in the cuDF environment is for developers who want IDE integration (usually for VS Code) with `black` or `clang-format`, which is harder to achieve when calling via pre-commit because it doesn't have the expected executable in the `PATH`.

If no other RAPIDS packages are using `mypy` and no developers want it installed in the development environment for purposes of IDE integration, it would be safe to remove from the cuDF conda environment file. Same for other "developer" tools like `flake8`, `black`, `pydocstyle`. I'd think this is unrelated / unaffected by the integration repo so that pinning could probably be deleted entirely.

**Comment by ajschmidt8** [comment]
@bdice, got it. we can remove those packages from `integration` if you'd like. But we have to be prepared to deal with any unexpected fallout if we do. Seems like the easiest path here would be to just continue to keep the versions in-sync until we truly deprecate that repository. But if you're prepared to address any potential fallout, then feel free to open a PR to remove them :slightly_smiling_face:.

**Comment by bdice** [comment]
> Seems like the easiest path here would be to just continue to keep the versions in-sync until we truly deprecate that repository.

I'm 100% in favor of that path. Just wanted to clarify that the integration `mypy` is not used by cuDF CI, so that we can revisit and simplify in the future.

File: .pre-commit-config.yaml
Comment (vyasr) [reply .pre-commit-config.yaml]: I'm late to the game here, and I'll follow up in more depth on #11661, but I agree with @wence- that this is trickier than it seems. `cupy` in particular would become a problem, since we've definitely had e.g. docs-only contributions before where users didn't even need to have CUDA or cudf-compatible GPU on their machine.

**Comment by vyasr** [comment]
I'm fine with the path outlined above. I agree that we can remove linters from the environment in the long run and just relyin on pre-commit, and I'm also happy to let this happen organically as integration is phased out rather than trying to be proactive in simplifying a repo that is going away anyway.

File: .pre-commit-config.yaml
Comment (vyasr) [inline .pre-commit-config.yaml:32]: In light of #11661 is this necessary, or can we wait to introduce this until we have a more holistic strategy for which dependencies get included here?

File: conda/environments/cudf_dev_cuda11.5.yml
Comment (vyasr) [inline conda/environments/cudf_dev_cuda11.5.yml:47]: Same question as above about whether we want this now.

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:110]: I'm a bit surprised that this logic works despite https://github.com/rapidsai/cudf/pull/11640#discussion_r961524007. I guess the parsing of the Python version from `sys.version_info` is something that mypy does support despite the static nature of the analysis?

File: python/cudf/cudf/core/indexed_frame.py
Comment (vyasr) [inline python/cudf/cudf/core/indexed_frame.py:63]: Rather than repeating this logic everywhere, let's move it to [`cudf._typing`](https://github.com/rapidsai/cudf/blob/branch-22.10/python/cudf/cudf/_typing.py).

File: setup.cfg
Comment (vyasr) [inline setup.cfg:41]:
```suggestion
# they are imported by any checked file.
```

File: setup.cfg
Comment (vyasr) [inline setup.cfg:50]: :vomiting_face:

**Review by vyasr — CHANGES_REQUESTED** [review]
Apologies for the delay in re-reviewing this. Looks mostly fine, there are a couple of minor questions I have left then we should be good to go.

File: .pre-commit-config.yaml
Comment (wence-) [reply .pre-commit-config.yaml:32]: This is a necessary dependency. Without it:
```
$ pre-commit run --all-files mypy
python/cudf/cudf/utils/cudautils.py:5: error: Library stubs not installed for "cachetools" (or incompatible with Python 3.9)
python/cudf/cudf/core/udf/utils.py:5: error: Library stubs not installed for "cachetools" (or incompatible with Python 3.9)
python/cudf/cudf/core/udf/utils.py:5: note: Hint: "python3 -m pip install types-cachetools"
python/cudf/cudf/core/udf/utils.py:5: note: (or run "mypy --install-types" to install all missing stub packages)
python/cudf/cudf/core/udf/utils.py:5: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

File: python/cudf/cudf/core/dataframe.py
Comment (wence-) [reply python/cudf/cudf/core/dataframe.py:110]: https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks

File: setup.cfg
Comment (wence-) [reply setup.cfg:41]: Went with "a".

File: python/cudf/cudf/core/indexed_frame.py
Comment (wence-) [reply python/cudf/cudf/core/indexed_frame.py]: Done.

**Comment by wence-** [comment]
> This will also need to be updated in the `integration` repository below. Can you update the `mypy` version and add the `types-cachetools` package to the `rapids-build-env` package?

Done here: https://github.com/rapidsai/integration/pull/531

File: python/cudf/cudf/core/dataframe.py
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py]: Nice that's cool.

**Review by vyasr — APPROVED** [review]

**Review by ajschmidt8 — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
